File: app/src/main/java/com/lfork/a98620/lfree/data/imdata/local/MessageLocalDataSource.java
package com.lfork.a98620.lfree.data.imdata.local;

import com.lfork.a98620.lfree.data.imdata.MessageDataSource;
import com.lfork.a98620.lfree.imservice.MessageListener;
import com.lfork.a98620.lfree.imservice.message.Message;
import com.lfork.a98620.lfree.imservice.message.MessageContentType;

import org.litepal.crud.DataSupport;
import org.litepal.crud.callback.FindMultiCallback;

import java.util.List;

public class MessageLocalDataSource implements MessageDataSource {

    private static MessageLocalDataSource INSTANCE;

    public static MessageLocalDataSource getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new MessageLocalDataSource();
        }
        return INSTANCE;
    }

    public static void destroyInstance() {
        INSTANCE = null;
    }

    @Override
    public void getMessages(int id, MessageContentType type, GeneralCallback<List<Message>> callback) {
        DataSupport.where("receiverid=? or senderid=?" ,id+"", id+"").order("id desc")
                .findAsync(Message.class)
                .listen(new FindMultiCallback() {
                    @Override
                    public <T> void onFinish(List<T> t) {

                        if (t.size() < 1) {
                            callback.failed("没有消息记录");
                        } else {
                            callback.succeed((List<Message>) t);
                        }
                    }
                });


    }

    @Override
    public void setMessageListener(MessageListener listener) {
        //消息监听交给 remote进行处理
    }


    @Override
    public void saveAndSendMessage(Message msg, GeneralCallback<Message> callback) {

        msg.saveAsync().listen(success -> callback.succeed(null));

    }

    @Override
    public void clearMessages(int id, MessageContentType type) {

    }

}

File: app/src/main/java/com/lfork/a98620/lfree/data/imdata/local/MessageLocalDataSource.kt
package com.lfork.a98620.lfree.data.imdata.local

import com.lfork.a98620.lfree.data.DataSource.GeneralCallback
import com.lfork.a98620.lfree.data.imdata.MessageDataSource
import com.lfork.a98620.lfree.imservice.MessageListener
import com.lfork.a98620.lfree.imservice.message.Message
import com.lfork.a98620.lfree.imservice.message.MessageContentType
import org.litepal.crud.DataSupport
import org.litepal.crud.callback.FindMultiCallback

/**
 *
 * Created by 98620 on 2018/9/3.
 */
class MessageLocalDataSource : MessageDataSource {

    companion object {
        private var INSTANCE: MessageLocalDataSource? = null
        fun getInstance(): MessageLocalDataSource {
            if (INSTANCE == null) {
                INSTANCE = MessageLocalDataSource()
            }
            return INSTANCE as MessageLocalDataSource
        }

        fun destroyInstance() {
            INSTANCE = null
        }
    }

    override fun getMessages(id: Int, type: MessageContentType, callback: GeneralCallback<List<Message>>) {
        DataSupport.where("receiverid=? or senderid=?", id.toString() + "", id.toString() + "").order("messageID")  //messageId 实际上是message生成的时间 System.currentTimeMillis()
                .findAsync(Message::class.java)
                .listen(object : FindMultiCallback {
                    override fun <T> onFinish(t: List<T>) {

                        if (t.size < 1) {
                            callback.failed("没有消息记录")
                        } else {
                            callback.succeed(t as List<Message>)
                        }
                    }
                })

    }

    override fun setMessageListener(listener: MessageListener) {
        //消息监听交给 remote进行处理
    }

    override fun saveAndSendMessage(msg: Message, callback: GeneralCallback<Message>) {
        msg.saveAsync().listen { success ->
            if (success) {
                callback.succeed(Message())
            } else {
                callback.failed("消息重复")
            }
        }

    }

    override fun clearMessages(id: Int, type: MessageContentType) {

    }


}

